### PR TITLE
validate various displayio args; new pin validation routines; don't use mp_const_none if NULL will do

### PR DESCRIPTION
**Pull request by dhalbert**
There was some missing pin argument validation in `displayio`, and when starting to fix that, I decided to simplify how pin argument validation and free pin checking was done.

- Replace `assert_pin` completely, and subsume most uses of `assert_free_pin()` by adding the new  routines  `validate_is_free_pin()` and `validate_is_free_pin_or_none()`, which check whether an `mp_obj_t` is a pin and return an `mcu_pin_obj_t *` if it is. This reduces several lines of code to one, improving readability and reducing boilerplate.
- Be consistent that optional `*_obj_t` arguments passed to `common_hal_...` (mostly pins) are passed as `NULL` instead of `mp_const_none`.  Also, use of `mp_const_none` in common-hal is minimized.

 The changes save about 260 bytes in the Trinket M0 build.

### CODE REVIEW

**Comment by tannewt** [comment]
I didn't look at the code (add me as a reviewer or @ me when ready) but I think `validate_pin_is_free` would read better.

**Comment by dhalbert** [comment]
> I didn't look at the code (add me as a reviewer or @ me when ready) but I think `validate_pin_is_free` would read better.

The old `assert_pin_free()` did not actually check that the object was a pin, so I was trying to make that clearer. But I agree the name was a bit awkward. I think maybe `validate_obj_is_pin()` and `validate_obj_is_free_pin()` would be more grammatical and clearer.

**Comment by tannewt** [comment]
`validate_obj_is_free_pin` works for me. Thanks!

**Comment by dhalbert** [comment]
The churn of trying `checkout@v2` (see #2683( and then reverting it fixed the build problems.

**Comment by dhalbert** [comment]
Now I am getting a "ERROR: THESE PACKAGES DO NOT MATCH THE HASHES FROM THE REQUIREMENTS FILE." errors on a few random builds, perhaps because of VM problems??

File: shared-bindings/displayio/ParallelBus.c [hunk 1]
```diff
@@ -76,18 +76,12 @@ STATIC mp_obj_t displayio_parallelbus_make_new(const mp_obj_type_t *type, size_t
...
-    assert_pin_free(chip_select);
-    assert_pin_free(write);
-    assert_pin_free(read);
-    assert_pin_free(reset);
+    mcu_pin_obj_t *data0 = validate_obj_is_free_pin(args[ARG_data0].u_obj);
+    mcu_pin_obj_t *command = validate_obj_is_free_pin(args[ARG_command].u_obj);
+    mcu_pin_obj_t *chip_select = validate_obj_is_free_pin(args[ARG_chip_select].u_obj);
+    mcu_pin_obj_t *write = validate_obj_is_free_pin(args[ARG_write].u_obj);
+    mcu_pin_obj_t *read = validate_obj_is_free_pin(args[ARG_read].u_obj);
+    mcu_pin_obj_t *reset = validate_obj_is_free_pin(args[ARG_reset].u_obj);
```
Comment (tannewt) [inline shared-bindings/displayio/ParallelBus.c]: I think reset should be required above.

**Comment by tannewt** [comment]
FYI this breaks Trinket M0 and likely any other boards that use a DotStar as the status LED. The SPI constructor call needs to be updated to pass NULL instead of mp_const_none here: https://github.com/adafruit/circuitpython/blob/master/supervisor/shared/rgb_led_status.c#L128

**Comment by dhalbert** [comment]
I looked through all the `mp_const_none` uses, but clearly not for all values of "all". :)

**Pull request merged**